### ISSUE TITLE
Possible Vulnerabilities

### ISSUE THREAD

**Issue by denny999222** [FivePixels/Tribble#3]
1. If the expected payment is $10.00, the user can send in $0.01 and put $10.00 inside the NOTES of the payment and it will still pass. So instead you can check the email header for the specific amount instead.

2. If someone sends a regular email with the keywords that is being checked for, it will still pass. So you can check if the email is actually coming from Venmo or PayPal or cash app

Some extra security checks can be check that the date of the email and the date of when the payment was sent matches up.

But regardless I do love the approach of accepting payments without a paywall on a discord bot. Exactly what I was looking for!

**Comment by FivePixels** [comment]
Thank you for the feedback and calling these out. I'll start implementing a fix for these ASAP.

**Comment by FivePixels** [comment]
After rereading (2), I don't believe its solvable by checking the header of the email. Do you mean the subject of the email? The issue with that is a user can exploit it still using the note so long as they know what the string is it uses to compare.

**Comment by FivePixels** [comment]
PR https://github.com/FivePixels/Tribble/pull/4 opened for a fix for (2).

**Comment by denny999222** [comment]
I'm actually doing multiple checks that should resolve most of the security issues. The key is checking data that the user cannot alter (this is still vulnerable), however the main security check really just checking if the email is actually coming from PayPal and Venmo

Venmo
1. Check email is coming from venmo@venmo.com in the "From" of the headers
2. Check "Subject" from header price matches expected price

Paypal
1. Check email is coming from service@paypal.com in the "From" of the headers
2. Check the email snippet portion (before the "notes") that the price matches expected price

I'm also checking for both that the ticketId is specifically inside the "NOTES" portion of the email and not anywhere else

**Comment by denny999222** [comment]
If the malicious user tries their hardest, I'm sure they'll be able to find a vulnerability, since the check is very weak since its only through email. Maybe email spoofing can still bypass, but I'm unsure

**Comment by denny999222** [comment]
There are many other ways in which you can decrease the risk of a vulnerability but no sure fire 100% method. One method I believe that can be very secure is you can change the primary email address that gets the email notifications from payments to an obscure email (E.g "abced2113VerysecureEmail.com"), and the user won't be able to figure out which email to send to.

**Comment by FivePixels** [comment]
> Check email is coming from [venmo@venmo.com](mailto:venmo@venmo.com) in the "From" of the headers
> Check email is coming from [service@paypal.com](mailto:service@paypal.com) in the "From" of the headers

[096a0be](https://github.com/FivePixels/Tribble/pull/4/commits/096a0be82f19db33e4eea2b00cd7794d576b1b71) addresses these two.  My next commit to the PR will add more information to the provider options for more reusability.

> Check "Subject" from header price matches expected price

I've got this implemented locally, will be pushing to the PR. Also have a check for malicious emails (in the case an email "coming from paypal or venmo" has the confirmation code in the subject.

> I'm also checking for both that the ticketId is specifically inside the "NOTES" portion of the email and not anywhere else

How are you checking this? I've decoded the base64 body and I'm getting the HTML of the PayPal email. I'm not sure how else to check this other than using regex to see if HTML before and after the code is the container for the note. 

> Check the email snippet portion (before the "notes") that the price matches expected price

I think if we are checking the body the snippet check isn't necessary, it would only be redundant. 

If you'd like, feel free to open a PR and contribute. If you've already worked on changes I don't see much of a point of me doing the same thing you've already done.
